zebra, bgpd: EVPN VXLAN Multihome support with extern mode

mrin-g opened the pull request:
Introduction of "--kernel-mac-ext-learn" mode of operation in zebra.
This delivers a comprehensive set of enhancements and fixes to support and validate EVPN VXLAN multihoming in external learning mode.

**EVPN Multihoming External Mode Support**:
- New option as  "--kernel-mac-ext-learn" in zebra startup.
- Both data plane and control plane MACs are marked with extern_learn, to prevent kernel aging.
- With the new 'protocol' field in bridge fdb, hardware learnt ('hw') and control plane learnt('zebra') are differentiated.
- 'extern_mode' handling of netlink messaging to debounce 'zebra' protocol messages and accept 'hw' mode for data plane learnt MAC
- 'ARP Suppression' presently not supported for 'extern only' mode of operation
- Specific handling of Sync MAC Update, Delete in extern mode as there is no kernel aging and thereby explicit cleanup is necessary.

**Test Enhancements and New Scenarios**:
The test framework is updated to cover EVPN VXLAN multihoming with both L2VNI and L3VNI, specifically targeting external learning mode. This includes new logic to handle and verify the correct operation of Ethernet Segments (ES), Designated Forwarder (DF) election, and VTEP peer relationships.
Added and updated topotests under bgp_evpn_mh_l2l3vni_ext_learn to: Validate correct ES discovery and advertisement for both local and remote PEs. Check VTEP peer lists for accuracy, including handling of downed VTEPs and ES state transitions. Ensure L2VNI and L3VNI are correctly instantiated and associated with the appropriate VRFs and VXLAN interfaces. Test orphaned hosts, dual-attached hosts, and single-attached hosts in various failure and recovery scenarios. Utility and Parser Functions:
Utility functions in lib/bgp_evpn.py(new) is added, such as a parser for the show evpn vni json command. This parser extracts VNI-to-type mappings, simplifying validation of VNI configuration and type (L2/L3) in automated tests. These changes adds the test coverage and reliability for EVPN VXLAN multihoming in external mode, making it easier to detect regressions and validate new features.

**Per File change summary**:

- bgpd/bgp_evpn.c:
   Addition of flog_err for error in installation of VNI MAC and IP to zebra. Non functional.

- include/linux/rtnetlink.h:
  RTPROT_HW definition as 'protocol' in bridge fdb

- zebra/debug_nl.c:
  Updated Netlink print infra to include 'protocol' field in netlink msg dump

- zebra/main.c:
  Add option '--kernel-ext-learn' in zebra startup.

- zebra/rt_netlink.c:
  Debounce/Accept MAC update from kernel based on 'protocol' field (i.e NDA_PROTOCOL)
  Add ext_flags e.g NTF_E_MH_PEER_SYNC as Peer Sync for 'extern_only' mode.
  Add NTF_EXT_LEARNED flag for MAC in dplane programming, also drop for netlink messages from kernel without the flag 

- zebra/zebra_evpn.c:
  ARP ND Suppression 'set' and 'get' function definition 

- zebra/zebra_evpn.h:
  Corresponding, ARP ND Suppression 'set' and 'get' function declaration 

- zebra/zebra_evpn_mac.c:
  zebra_evpn_sync_mac_dp_install: For Sync MAC DPLANE installation, make sure the MAC is inactive for extern mode
  In zebra_evpn_mac_hold_exp_cb: In extern only mode, explicit flush of the mac is required
  In zebra_evpn_sync_mac_del: In extern mode, only reprogram MACif it does not have any PEER flags or neighbor and LOCAL INACTIVE. This helps to clear up the "static" for PEER_PROXY on proxy withdrawal

- zebra/zebra_evpn_mh.c:
  ES Info Set: In extern only mode, use the API zebra_evpn_es_bypass_update_macs to clean MAC from interfaces and ES 
 
- zebra/zebra_evpn_mh.h:
  Declare API zebra_evpn_flush_local_mac as its used in zebra_evpn_mac.c for hold timer expiry cleanup and sync delete of MAC 

- zebra/zebra_evpn_neigh.c:
  Prevent neighbor entry installation n DPLANE for remote neighbor if ARP/ND suppression is not enabled. For Extern only mode, ARP/ND Supression is not enabled

- zebra/zebra_router.c:
   For extern only mode, disable ARP suppression

-  zebra/zebra_router.h:
   new bool field 'kernel_mac_ext_learn' in zebra_router, to store extern only mode 

- zebra/zebra_vxlan.c:
   Add print for EVPN ARP/ND Suppression in "show evpn" 

-  zebra/zebra_vxlan.h:
   Accessor for Extern Only mode i.e zebra_mac_ext_learn_mode

Review for Kernel Change for Protocol field Addition: https://lore.kernel.org/netdev/20250818175258.275997-1-mrghosh@cisco.com/T/#u
Review for iproute2 change for protocol field Addition: https://lore.kernel.org/netdev/20250818193756.277327-1-mrghosh@cisco.com/T/#u

mrin-g committed:
zebra : EVPN VXLAN Multihome extern mode: Enable --kernel-mac-ext-learn

Introduction of "--kernel-mac-ext-learn" mode of operation in zebra.

This is part of the change is to enable the
 option '--kernel-mac-ext-learn' in zebra.

Rationale:
EVPN Multihoming External Mode Support is to enable platforms doing
Hardware Based MAC Learning and Aging to support EVPN VXLAN Multihome.

MAC's in this mode for both data and control plane will be
 marked and programmed as 'extern_only' in kernel, so that kernel aging
 is disabled for these MACs. Zebra along with Hardware(HW) will control
 these MACs for control plane and data plane respectively.

Per File change summary:
zebra/main.c:
  Add option '--kernel-mac-ext-learn' in zebra startup.
  Alignment changes for the other options to match the new option width
  and keep description alignment consistent
zebra/zebra_router.c:
   - zebra_router_init definition updation to pass
      kernel_mac_ext_learn mode
   - store kernel_mac_ext_learn in zrouter zav structure
zebra/zebra_router.h:
   - new bool field 'kernel_mac_ext_learn' in zebra_router, to
     store extern only mode
zebra/zebra_vxlan.h:
   - Accessor for Extern Only mode i.e zebra_mac_ext_learn_mode

Signed-off-by: Mrinmoy Ghosh <mrinmoy_g@hotmail.com>
Signed-off-by: Mrinmoy Ghosh <mrghosh@cisco.com>
Signed-off-by: Patrice Brissette <pbrisset@cisco.com>

donaldsharp commented on tests/topotests/bgp_evpn_mh_l2l3vni_ext_learn/spine1/evpn.conf:
```diff
@@ -0,0 +1,18 @@
+frr defaults datacenter
```
I see the addition of a bunch of empty files this makes no sense.  Please remove them.  The topotest should be it's own commit too.

donaldsharp replied:
We also require all new tests to have a integrated config.  Please change it.

Commit:
zebra: New protocol RTPROT_HW for hardware learnt MAC

Protocol field is added in bridge FDB, to distinguish between
 MAC addresses learned via the control plane and those learned
 via the data plane with hardware aging.

Protocol 'hw' (i.e RTPROT_HW aka hardware) for MAC learnt by hardware
 will be used for data plane(hardware) learnt MAC while
 existing protocol 'zebra' to be used for control plane learnt ones.

Kernel Patch in review:
https://lore.kernel.org/netdev/20250818175258.275997-1-mrghosh@cisco.com/
iproute2 patch in review:
https://lore.kernel.org/netdev/20250818193756.277327-1-mrghosh@cisco.com/

Signed-off-by: Mrinmoy Ghosh <mrinmoy_g@hotmail.com>
Signed-off-by: Mrinmoy Ghosh <mrghosh@cisco.com>
Signed-off-by: Patrice Brissette <pbrisset@cisco.com>

Commit:
doc: zebra: Doc update for zebra option kernel-mac-ext-learn

Doc update for new zebra option '--kernel-mac-ext-learn'

Signed-off-by: Mrinmoy Ghosh <mrinmoy_g@hotmail.com>
Signed-off-by: Mrinmoy Ghosh <mrghosh@cisco.com>

mrin-g committed:
zebra : EVPN VXLAN MH extern mode: Update debug print for NDA_PROTOCOL

This is to update netlink debug print,
 to print 'protocol' i.e NDA_PROTOCOL field for netlink messages.

Signed-off-by: Mrinmoy Ghosh <mrinmoy_g@hotmail.com>
Signed-off-by: Mrinmoy Ghosh <mrghosh@cisco.com>

mrin-g committed:
zebra: ARP/ND Suppression disabled for kernel-mac-ext-learn mode

For MAC External mode, presently ARP/ND Suppression is not supported,
 thereby disabled. Remote Neighbor programming is also not required,
 based on zebra_mac_ext_learn_mode.

Signed-off-by: Mrinmoy Ghosh <mrinmoy_g@hotmail.com>
Signed-off-by: Mrinmoy Ghosh <mrghosh@cisco.com>
Signed-off-by: Patrice Brissette <pbrisset@cisco.com>

mrin-g committed:
zebra: netlink message handling for mac_extern_learn mode

Receive handling:
- Debounce of AF_BRIDGE netlink message based on 'protocol' field ZEBRA
- Update zlog to include protocol
- Ignore feedback netlink from zebra based on protocol zebra
- Ignore if "NTF_EXT_LEARNED" flag is not marked in the netlink message
  for extern only mode of operation
Send Operation:
- Always mark MAC as NTF_EXT_LEARNED in extern only mode
- Mark Sync MAC as NTF_E_MH_PEER_SYNC, only in extern only mode
- Peer-Sync flag (i.e NTF_E_MH_PEER_SYNC) state print in Tx zlog

Signed-off-by: Mrinmoy Ghosh <mrinmoy_g@hotmail.com>
Signed-off-by: Mrinmoy Ghosh <mrghosh@cisco.com>
Signed-off-by: Patrice Brissette <pbrisset@cisco.com>

mrin-g committed:
tests: topotests: EVPN VXLAN MH extern learn mode tests

Topotests added under bgp_evpn_mh_l2l3vni_ext_learn to:
- Validate correct ES discovery and advertisement for both local and
 remote PEs.
- Check VTEP peer lists for accuracy, including handling of downed VTEPs
  and ES state transitions.
- Ensure L2VNI and L3VNI are correctly instantiated and associated with the
 appropriate VRFs and VXLAN interfaces.
- Test orphaned hosts, dual-attached hosts, and single-attached hosts
 in various failure and recovery scenarios.
- MAC 'protocol' state transitions i.e data plane learnt to
  control plane learnt and vice versa, delete, relearn in both peers
  sequeunces
Utility and Parser Functions:
Utility functions in lib/bgp_evpn.py(new) are added.
These changes adds the test coverage and reliability for EVPN VXLAN
 multihoming in external mode, making it easier to detect regressions and
 validate new features.

Signed-off-by: Mrinmoy Ghosh <mrinmoy_g@hotmail.com>
Signed-off-by: Mrinmoy Ghosh <mrghosh@cisco.com>
Signed-off-by: Patrice Brissette <pbrisset@cisco.com>

mrin-g committed:
bgpd: install evpn failure logging

flog_err added for evpn install to zebra failure for
 ip and mac.

Signed-off-by: Mrinmoy Ghosh <mrinmoy_g@hotmail.com>
Signed-off-by: Mrinmoy Ghosh <mrghosh@cisco.com>
Signed-off-by: Mike Mallin <mmallin@cisco.com>

mrinmoyg commented:
> This commit needs to be broken up into small logical units of work such that a reviewer can look at it and understand the logical steps that get you to this new feature. I have given some suggestions on how to do this already. I am sure that there are other places it can be broken up some.
> 
> Additionally I see no documentation at all. This needs to be added too. I can review further once this is broken up

Its broken down now, with the zebra.rst file updated for the documentation on the new option.

louis-6wind commented on tests/topotests/bgp_evpn_mh_l2l3vni_ext_learn/test_evpn_mh_l2l3vni_ext_learn.py:
```diff
@@ -0,0 +1,1047 @@
+    )
+
+    # Wait for sync
+    time.sleep(5)
```
same remark

louis-6wind reviewed (changes requested):
a few comments, on the form for the moment

ton31337 commented on tests/topotests/bgp_evpn_mh_l2l3vni_ext_learn/test_evpn_mh_l2l3vni_ext_learn.py:
```diff
@@ -0,0 +1,1047 @@
+
+    # Verify MAC is removed from torm11's bridge FDB, but would be added back
+    test_fn = partial(bgp_evpn.check_mac_in_bridge, dut, mac, dev, vlan, proto="zebra")
+    _, result = topotest.run_and_expect(test_fn, None, count=10, wait=1)
```
Use higher values, 15 seconds is the minimum.

ton31337 commented on tests/topotests/bgp_evpn_mh_l2l3vni_ext_learn/test_evpn_mh_l2l3vni_ext_learn.py:
```diff
@@ -0,0 +1,1047 @@
+    test_fn = partial(
+        bgp_evpn.check_mac_in_bridge, dut2, mac, dev2, vlan, proto="zebra"
+    )
+    _, result = topotest.run_and_expect(test_fn, None, count=10, wait=1)
```
Ditto.

ton31337 commented on tests/topotests/bgp_evpn_mh_l2l3vni_ext_learn/test_evpn_mh_l2l3vni_ext_learn.py:
```diff
@@ -0,0 +1,1047 @@
+
+    # Check torm12 now has proto hw
+    test_fn = partial(bgp_evpn.check_mac_in_bridge, dut2, mac, dev2, vlan, proto="hw")
+    _, result = topotest.run_and_expect(test_fn, None, count=10, wait=1)
```
Ditto.

ton31337 commented on tests/topotests/bgp_evpn_mh_l2l3vni_ext_learn/test_evpn_mh_l2l3vni_ext_learn.py:
```diff
@@ -0,0 +1,1047 @@
+
+    # Check torm11 still has proto zebra
+    test_fn = partial(bgp_evpn.check_mac_in_bridge, dut, mac, dev, vlan, proto="zebra")
+    _, result = topotest.run_and_expect(test_fn, None, count=10, wait=1)
```
Ditto.

ton31337 commented on tests/topotests/bgp_evpn_mh_l2l3vni_ext_learn/test_evpn_mh_l2l3vni_ext_learn.py:
```diff
@@ -0,0 +1,1047 @@
+
+    # Verify proto=hw on torm11
+    test_fn = partial(bgp_evpn.check_mac_in_bridge, dut, mac, dev, vlan, proto="hw")
+    _, result = topotest.run_and_expect(test_fn, None, count=10, wait=1)
```
Ditto.

ton31337 commented on zebra/rt_netlink.c:
```diff
@@ -4226,16 +4235,39 @@ static int netlink_macfdb_change(struct nlmsghdr *h, int len, ns_id_t ns_id)
+			if (!(ndm->ndm_flags & NTF_EXT_LEARNED)) {
+				if (IS_ZEBRA_DEBUG_KERNEL)
+					zlog_debug("        Should drop entry due to missing extern learn, FLAGS = 0x%x",
+						   ndm->ndm_flags & NTF_EXT_LEARNED);
```
If we want to print all flags `FLAGS = `, then why we expect only `NTF_EXT_LEARNED`?

ton31337 commented on zebra/zebra_evpn_mac.c:
```diff
@@ -1588,7 +1596,15 @@ void zebra_evpn_sync_mac_del(struct zebra_mac *mac)
-	if (old_static != new_static)
+	if (zebra_mac_ext_learn_mode() && !new_static &&
+	    CHECK_FLAG(mac->flags, ZEBRA_MAC_LOCAL_INACTIVE)) {
+		// Clear the MAC from Peer Proxy, as no age out will happen for extern mode
```
Let's drop this comment or replace to `/* */`.

ton31337 commented on zebra/zebra_vxlan.c:
```diff
@@ -3905,6 +3905,8 @@ void zebra_vxlan_print_evpn(struct vty *vty, bool uj)
 		json_object_boolean_add(json, "isDetectionFreeze",
 					zvrf->dad_freeze);
+		json_object_string_add(json, "arpNdSuppress",
+				       zebra_evpn_get_arp_nd_suppress() ? "Yes" : "No");
```
Why not json_object_boolean_add?

ton31337 commented on doc/user/zebra.rst:
```diff
@@ -103,6 +103,14 @@ Besides the common invocation options (:ref:`common-invocation-options`), the
 
+.. option:: --kernel-ext-learn
+
+   Signal to zebra that its operating in kernel external learn mode.
```
I expect much broader documentation in this case, because as an operator, it's not clear when we should use it.

mrinmoyg commented on zebra/zebra_evpn.h:
```diff
@@ -36,6 +36,10 @@ struct zebra_evpn_show {
 
+/* Function to control ARP/ND suppression */
+void zebra_evpn_set_arp_nd_suppress(bool set_default);
+bool zebra_evpn_get_arp_nd_suppress(void);
```
You are right, ideally kernel extern learn and ARP ND Suppression support are orthogonal. However, presently we dont support ARP/ND Suppression with kernel extern learn mode i.e HW learn mode, hence those are linked. We were considering to add a separate CLI later which will invoke _zebra_evpn_set_arp_nd_suppress_ to disable it.
So intention of this function is to keep the ARP/ND suppression functionality separate from "extern learn mode".

mrin-g replied:
Some context for this approach:
Presently ARP/ND Suppression with Asymmetric IRB, is enabled by default. However with MAC Extern mode, we wont support it presently (IP mobility issues with different MAC), hence setting _zebra_evpn_set_arp_nd_suppress_ to false with this mode.
Here the functions are just getters and setters for the arp/nd suppression modes as that would be needed to program the remote neighbor. For symmetric IRB, remote neighbor programming is skipped, which is the only functional usage of  _zebra_evpn_get_arp_nd_suppress_ ; Which is not ideally related to mac extern mode.

louis-6wind replied:
My concern is that the functions let us think that we are setting bridge interface flag neigh_suppress (aka. ARP/NDP) but you are only setting a global variable zebra_evpn_do_arp_nd_suppress.

Since `kernel_mac_ext_learn` is always `!zebra_evpn_do_arp_nd_suppress`, it is useless to have such a misleading complexity.

You can instead test `kernel_mac_ext_learn` and put in comment the ARP/ND Suppression issue.

louis-6wind replied:
> However with MAC Extern mode, we wont support it presently (IP mobility issues with different MAC)
Are you referring to this issue https://github.com/FRRouting/frr/issues/14156 ?

mrin-g replied:
exactly (thanks for pointing it out!), in addition to host mobility, asymmetric IRB also has scaling issues as adjacencies are spread across the fabric. However thats a choice we wish to keep, as in some deployments its critical to have that. Presently for mac-extern, we wish to keep it symmetric. 
Anyways its just modularity paradigm; I will remove it to avoid confusion, adding the needful comments

mrin-g commented on zebra/zebra_evpn_mh.c:
```diff
@@ -2321,7 +2322,14 @@ static void zebra_evpn_es_local_info_set(struct zebra_evpn_es *es,
+		/* Note: though bypass route is called, but its effectively clearing all
+		 * macs
+		 */
+		zebra_evpn_es_bypass_update_macs(es, zif->ifp, false);
```
This is existing function, just leveraging https://wwwin-github.cisco.com/whitebox/frr/blob/b2d316eb00bc2e626e76ac382a1113405dc88df6/zebra/zebra_evpn_mh.c#L2680

mrin-g commented on zebra/zebra_evpn_mac.c:
```diff
@@ -1519,16 +1520,23 @@ static void zebra_evpn_mac_hold_exp_cb(struct event *t)
+		struct interface *ifp;
+		/* Upon expiry, MAC is delete from DP. */
+		zebra_evpn_mac_get_access_info(mac, &ifp, &vid);
+		zebra_evpn_flush_local_mac(mac, ifp);
```
`zebra_evpn_sync_mac_dp_install` does not trigger mac delete. It does remote mac delete if the ES interface is down with fast failover. However this is to cleanup the Local interface stamped MACs, agnostic to interface state. On a Hold down timer expiry, if hte MAC has not been relearnt this is to cleanup the kernel and data plane.

mrin-g commented on zebra/main.c:
```diff
@@ -442,6 +446,9 @@ int main(int argc, char **argv)
 		case 'R':
 			rt_table_main_id = atoi(optarg);
 			break;
+		case OPTION_KERNEL_MAC_EXT_LEARN:
```
Yes, thats a good point. Presently if we see there is a mismatch between the mac_external mode and the kernel  MAC update e.g https://github.com/FRRouting/frr/pull/19438/files#diff-c5ead220153b4b424c817d54e28ca166882af2a78e0958c106b585bf1ec652ecR4176 , we drop the entry. So no local MAC learning will happen without external flag in mac external mode. 
But presently its all config based and no operational mode is maintained, however the only tricky part to this dynamic mode, is if we do the evaluation per MAC basis; any bug in the underlying mac learning infra can cause flip flop between the modes. Would love to discuss your idea in a huddle.

donaldsharp commented on tests/topotests/bgp_evpn_mh_l2l3vni_ext_learn/test_evpn_mh_l2l3vni_ext_learn.py:
```diff
@@ -0,0 +1,1035 @@
+    host.run("ip link set dev %s address %s" % (bond_name, host_mac))
+
+
+def config_hosts(tgen, hosts):
```
yep cut-n-pasted

donaldsharp commented on zebra/rt_netlink.c:
```diff
@@ -4350,6 +4364,7 @@ ssize_t netlink_macfdb_update_ctx(struct zebra_dplane_ctx *ctx, void *data,
 	uint32_t update_flags;
 	bool nfy = false;
 	uint8_t nfy_flags = 0;
+	uint32_t nda_ext_flags = 0;
```
In this function nda_ext_flags are being set and debugged, but on line 4472 when it is time to encode these flags are not sent to the kernel.  Is this intentional?

donaldsharp commented on zebra/zebra_evpn_mh.c:
```diff
@@ -2101,8 +2103,7 @@ static void zebra_evpn_es_setup_evis(struct zebra_evpn_es *es)
 
-static void zebra_evpn_flush_local_mac(struct zebra_mac *mac,
-				       struct interface *ifp)
+void zebra_evpn_flush_local_mac(struct zebra_mac *mac, struct interface *ifp)
```
So both places where this function is called call zebra_evpn_mac_get_access_info where it is possible that a NULL pointer is returned for the ifp.  Can we put a guard in place to prevent that?  Is it actually possible for this to happen?

mjstapp commented:
Mrinmoy has moved to a different group; a new version of this EVPN-MH work is PR #21863
